Machine Learning/Regression Loss Functions/regression-loss-functions.ipynb

```diff
@@ -0,0 +1,415 @@
+    "axMiddle.set_ylim(800,1000)\n",
+    "axBottom.set_ylim(-20,75)\n",
+    "axMiddle.legend()\n",
+    "\n",
+    "pl.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that I broke the y-axis in the bottom diagrams to be able to see those small residuals!\n",
+    "\n",
+    "First, MSE tends to result in larger errors for small target values despite the fact that most of the targets are rather small. It rather focuses on producing smaller errors for large target values.\n",
+    "\n",
+    "Second, if we look at the density of the target values (top diagram), we see that qualitatively the \"transition point\" happens somewhere around 10: we have a high concentration of small targets <10 and low number of targets >>10. Quantitatively, this \"transition point\" is somewhere between 80th and 90th percentile. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Percentile 0\tvalue 0.00\n",
+      "Percentile 10\tvalue 0.06\n",
+      "Percentile 20\tvalue 0.15\n",
+      "Percentile 30\tvalue 0.31\n",
+      "Percentile 40\tvalue 0.54\n",
+      "Percentile 50\tvalue 0.96\n",
+      "Percentile 60\tvalue 1.74\n",
+      "Percentile 70\tvalue 3.10\n",
+      "Percentile 80\tvalue 6.18\n",
+      "Percentile 90\tvalue 16.46\n",
+      "Percentile 100\tvalue 4891.30\n"
+     ]
+    }
+   ],
+   "source": [
+    "quantiles = np.linspace(0, 1.0, 11)\n",
+    "qvalues = np.quantile(yexp, np.linspace(0, 1.0, 11))\n",
+    "for q, val in zip(quantiles, qvalues):\n",
+    "    print(\"Percentile {:.0f}\\tvalue {:.2f}\".format(q*100, val))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Setting the offset parameter to 1.0 or 100.0 undershoots and overshoots this \"transition point\". Both lead to similar results with greater residuals for small targets and lower residuals for large targets compared to the offset parameter of 10.0. However, trying to \"nail\" the correct transition point super exactly is not worth it: setting offset parameter to 6.0 or 16.0  won't change the results significantly compared to 10.0.\n",
+    "\n",
+    "### Conclusion\n",
+    "\n",
+    "To use MSLE with a certain parameter offset $t$ in your data frame analytics regression job, you can specify parameter `loss_function` (set it to `msle`) and `loss_function_parameter`. Try it, if you have one of the [numerous examples](https://en.wikipedia.org/wiki/Log-normal_distribution#Occurrence_and_applications) of positive target values which scale over a large scale. You can leave the `loss_function_parameter` as it is, but if you want to experiment with different offset values, we recommend to set the value somewhere at the 80th percentile of the target distribution and then increase by an order of magnitude and look at the results qualitatively."
```

Comment:
This is good! It might also be worth saying that there should be strong evidence for positive skew: Huber would do better if there are symmetric heavy tails. For example, either measure the skewness or plot the histogram and check for the heavy tail. This can also be done after fitting the model with MSE: for example, checking non-normality of the prediction errors.

Reply:
Updated